<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>GLMM with negative binomial distribution for sequencing count data — glmmSeq • glmmSeq</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="GLMM with negative binomial distribution for sequencing count data — glmmSeq" />
<meta property="og:description" content="Fits many generalised linear mixed effects models (GLMM) with negative
binomial distribution for analysis of overdispersed count data with random
effects. Designed for longitudinal analysis of RNA-Sequencing count data.
Wald type 2 Chi-squared test is used to calculate p-values." />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">glmmSeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="../articles/glmmSeq.html">
    <span class="fas fa-book"></span>
     
    Get Started
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fas fa-newspaper"></span>
     
    News
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="far fa-file-code"></span>
     
    Reference Manual
  </a>
</li>
<li>
  <a href="https://github.com/KatrionaGoldmann/glmmSeq">
    <span class="fab fa-github"></span>
     
    Github
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>GLMM with negative binomial distribution for sequencing count data</h1>
    <small class="dont-index">Source: <a href='https://github.com/KatrionaGoldmann/glmmSeq/blob/master/R/glmmSeq.R'><code>R/glmmSeq.R</code></a></small>
    <div class="hidden name"><code>glmmSeq.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Fits many generalised linear mixed effects models (GLMM) with negative
binomial distribution for analysis of overdispersed count data with random
effects. Designed for longitudinal analysis of RNA-Sequencing count data.
Wald type 2 Chi-squared test is used to calculate p-values.</p>
    </div>

    <pre class="usage"><span class='fu'>glmmSeq</span><span class='op'>(</span>
  <span class='va'>modelFormula</span>,
  <span class='va'>countdata</span>,
  <span class='va'>metadata</span>,
  id <span class='op'>=</span> <span class='cn'>NULL</span>,
  dispersion <span class='op'>=</span> <span class='cn'>NA</span>,
  sizeFactors <span class='op'>=</span> <span class='cn'>NULL</span>,
  reducedFormula <span class='op'>=</span> <span class='st'>""</span>,
  modelData <span class='op'>=</span> <span class='cn'>NULL</span>,
  designMatrix <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"lme4"</span>, <span class='st'>"glmmTMB"</span><span class='op'>)</span>,
  control <span class='op'>=</span> <span class='cn'>NULL</span>,
  family <span class='op'>=</span> <span class='va'>nbinom2</span>,
  cores <span class='op'>=</span> <span class='fl'>1</span>,
  removeSingles <span class='op'>=</span> <span class='cn'>FALSE</span>,
  zeroCount <span class='op'>=</span> <span class='fl'>0.125</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>,
  returnList <span class='op'>=</span> <span class='cn'>FALSE</span>,
  progress <span class='op'>=</span> <span class='cn'>FALSE</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>modelFormula</th>
      <td><p>the model formula. This must be of the form <code>"~ ..."</code>
where the structure is assumed to be <code>"counts ~ ..."</code>. The formula must
include a random effects term. For more information on formula structure
for random effects see <code><a href='https://rdrr.io/pkg/lme4/man/glmer.html'>lme4::glmer()</a></code></p></td>
    </tr>
    <tr>
      <th>countdata</th>
      <td><p>the sequencing count data matrix with genes in rows and
samples in columns</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p>a dataframe of sample information with variables in columns
and samples in rows</p></td>
    </tr>
    <tr>
      <th>id</th>
      <td><p>Optional. Used to specify the column in metadata which contains the
sample IDs to be used in repeated samples for random effects. If not
specified, the function defaults to using the variable after the "|" in the
random effects term in the formula.</p></td>
    </tr>
    <tr>
      <th>dispersion</th>
      <td><p>a numeric vector of gene dispersion. Not required for
<code>glmmTMB</code> models, as these determine and fit dispersion for each gene.</p></td>
    </tr>
    <tr>
      <th>sizeFactors</th>
      <td><p>size factors (default = NULL). If provided the <code>glmer</code>
offset is set to log(sizeFactors). For more information see``
<code><a href='https://rdrr.io/pkg/lme4/man/glmer.html'>lme4::glmer()</a></code></p></td>
    </tr>
    <tr>
      <th>reducedFormula</th>
      <td><p>Reduced design formula (default = "")</p></td>
    </tr>
    <tr>
      <th>modelData</th>
      <td><p>Optional dataframe. Default is generated by call to
<code>expand.grid</code> using levels of variables in the formula. Used to calculate
model predictions (estimated means &amp; 95% CI) for plotting via <a href='modelPlot.html'>modelPlot</a>.
It can therefore be used to add/remove points in <a href='modelPlot.html'>modelPlot</a>.</p></td>
    </tr>
    <tr>
      <th>designMatrix</th>
      <td><p>custom design matrix, used only for prediction</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>Specifies which package to use for fitting GLMM models. Either
"lme4" or "glmmTMB" depending on whether to use <a href='https://rdrr.io/pkg/lme4/man/glmer.html'>lme4::glmer</a> or
<a href='https://rdrr.io/pkg/glmmTMB/man/glmmTMB.html'>glmmTMB::glmmTMB</a> to fit GLMM models.</p></td>
    </tr>
    <tr>
      <th>control</th>
      <td><p>the <code>glmer</code> optimizer control. If <code>method = "lme4"</code> default is
<code>glmerControl(optimizer = "bobyqa")</code>). If
<code>method = "glmmTMB"</code> default is <code>glmmTMBControl()</code></p></td>
    </tr>
    <tr>
      <th>family</th>
      <td><p>Only used with <code>glmmTMB</code> models. Default is <code>nbinom2</code>. See
<a href='https://rdrr.io/pkg/glmmTMB/man/nbinom2.html'>glmmTMB::nbinom2</a></p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>number of cores to use. Default = 1.</p></td>
    </tr>
    <tr>
      <th>removeSingles</th>
      <td><p>whether to remove individuals without repeated measures
(default = FALSE)</p></td>
    </tr>
    <tr>
      <th>zeroCount</th>
      <td><p>numerical value to offset zeroes for the purpose of log
(default = 0.125)</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Logical whether to display messaging (default = TRUE)</p></td>
    </tr>
    <tr>
      <th>returnList</th>
      <td><p>Logical whether to return results as a list or <code>glmmSeq</code>
object (default = FALSE). Useful for debugging.</p></td>
    </tr>
    <tr>
      <th>progress</th>
      <td><p>Logical whether to display a progress bar</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Other parameters to pass to
<code><a href='https://rdrr.io/pkg/lme4/man/glmer.html'>lme4::glmer()</a></code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns an S4 class <code>GlmmSeq</code> object with results for gene-wise
general linear mixed models. A list of results is returned if <code>returnList</code>
is <code>TRUE</code> which is useful for debugging.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function is a wrapper for <code><a href='https://rdrr.io/pkg/lme4/man/glmer.html'>lme4::glmer()</a></code>. Wald type 2 Chi-squared test
is calculated as per <code><a href='https://rdrr.io/pkg/car/man/Anova.html'>car::Anova()</a></code> optimised for speed. Parallelisation is
provided using <a href='https://rdrr.io/r/parallel/mclapply.html'>parallel::mclapply</a> on Unix/Mac or <a href='https://rdrr.io/r/parallel/clusterApply.html'>parallel::parLapply</a> on
PC.</p>
<p>Setting <code>method = "glmmTMB"</code> enables an alternative method of fitting GLMM
using the <code>glmmTMB</code> package. This gives access to a variety of alternative
GLM family functions. Note, <code>glmmTMB</code> negative binomial models are
substantially slower to fit than <code>glmer</code> models with known dispersion due to
the extra time taken by <code>glmmTMB</code> to optimise the dispersion parameter.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='https://rdrr.io/pkg/lme4/man/glmer.html'>lme4::glmer</a> <a href='https://rdrr.io/pkg/lme4/man/lmerControl.html'>lme4::glmerControl</a> <a href='https://rdrr.io/pkg/glmmTMB/man/glmmTMB.html'>glmmTMB::glmmTMB</a>
<a href='https://rdrr.io/pkg/glmmTMB/man/nbinom2.html'>glmmTMB::nbinom2</a> <a href='https://rdrr.io/pkg/glmmTMB/man/glmmTMBControl.html'>glmmTMB::glmmTMBControl</a> <a href='https://rdrr.io/pkg/car/man/Anova.html'>car::Anova</a></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='r-in'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>PEAC_minimal_load</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>disp</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>tpm</span>, <span class='fl'>1</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='op'>{</span></span>
<span class='r-in'><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/cor.html'>var</a></span><span class='op'>(</span><span class='va'>x</span>, na.rm <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/mean.html'>mean</a></span><span class='op'>(</span><span class='va'>x</span>, na.rm <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span><span class='op'>/</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/mean.html'>mean</a></span><span class='op'>(</span><span class='va'>x</span>, na.rm <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>**</span><span class='fl'>2</span><span class='op'>)</span></span>
<span class='r-in'><span class='op'>}</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>MS4A1glmm</span> <span class='op'>&lt;-</span> <span class='fu'>glmmSeq</span><span class='op'>(</span><span class='op'>~</span> <span class='va'>Timepoint</span> <span class='op'>*</span> <span class='va'>EULAR_6m</span> <span class='op'>+</span> <span class='op'>(</span><span class='fl'>1</span> <span class='op'>|</span> <span class='va'>PATID</span><span class='op'>)</span>,</span>
<span class='r-in'>                     countdata <span class='op'>=</span> <span class='va'>tpm</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>2</span>, <span class='op'>]</span>,</span>
<span class='r-in'>                     metadata <span class='op'>=</span> <span class='va'>metadata</span>,</span>
<span class='r-in'>                     dispersion <span class='op'>=</span> <span class='va'>disp</span>,</span>
<span class='r-in'>                     verbose <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/attributes.html'>attributes</a></span><span class='op'>(</span><span class='va'>MS4A1glmm</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>  [1] "info"           "formula"        "stats"          "predict"       </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>  [5] "reducedFormula" "countdata"      "metadata"       "modelData"     </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>  [9] "optInfo"        "errors"         "vars"           "class"         </span>
<span class='r-in'></span>
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.2/css/all.css">

<style>
  #left {
    float:left;
  }
  #right {
    width:100%;
    position:relative;
    text-align:right;
    align:right;
    float:right;
  }
  *.icon-grey {color: lightgrey}
  j {
    text-align:center;
    vertical-align:middle;
    position: relative;
    padding-bottom: 12px;
  }
  .smadge:after{
    content:"EMR";
    position: absolute;
    background: lightgrey;
    height:1.2rem;
    top:1rem;
    right:-1rem;
    width:2.3rem;
    text-align: center;
    line-height: 1rem;;
    font-size: 1rem;
    color: #ff3333;
    border-radius: 10%;
  }
  .bladge:after{
    content:"C4TB";
    position: absolute;
    background: lightgrey;
    height:1.2rem;
    top:1rem;
    right:-1rem;
    width:2.7rem;
    text-align: center;
    line-height: 1rem;;
    font-size: 1rem;
    color: #ff3333;
    border-radius: 10%;
  }
</style>

<div id="left">
Developed by the <a href="https://www.qmul.ac.uk/whri/emr/">Centre for Experimental Medicine & Rheumatology</a>, QMUL
</div>

<div id="right">
<a href="https://www.qmul.ac.uk/whri/emr/"><j class="fa fa-university fa-2x  icon-grey smadge" aria-hidden="false" style="color:#ff3333"></j></a>
<span style="margin-left:1rem"></span>

<a href="https://www.qmul.ac.uk/c4tb/"><j class="fa fa-university fa-2x  icon-grey bladge" aria-hidden="false" style="color:#ff3333"></j></a>
<span style="margin-left:1rem"></span>

<a href="mailto:k.goldmann@qmul.ac.uk" class="fas fa-envelope fa-2x" style="color:#ff3333">
</a><span style="margin-left:1rem"></span>

<a href="https://github.com/KatrionaGoldmann"><i class="fab fa-github fa-2x" style="color:#ff3333"></i></a>
<span style="margin-left:1rem"></span>

<a href="https://twitter.com/EMR_QMUL"><i class="fab fa-twitter fa-2x" style="color:#ff3333"></i></a>
</div>

      </footer>
   </div>

  


  </body>
</html>


